common/simple_dom.h

```diff
@@ -0,0 +1,274 @@
+};  rstring_view32 _key, _value; // of _key and _value of rstring_view)
+
+    friend struct NodeWrapper;
+
+public:
+    virtual size_t num_children() const __attribute__((pure)) = 0;
+
+    // get the i-th child node
+    virtual Node* get(size_t i) const __attribute__((pure)) = 0;
+
+    // get the first child node with a specified `key`
+    virtual Node* get(str key) const __attribute__((pure)) = 0;
+
+    RootNode* root() const __attribute__((pure)) {
+        auto ptr = this;
+        while (ptr->_parent)
+            ptr = ptr->_parent;
+        return (RootNode*)ptr;
+    }
+
+    NodeWrapper wrapper();
+};
+
+class RootNode : public Node {
+protected:
+    RootNode() = default;
+
+    // reference counting is only applied to root
+    // node, which represents the whole document
+    std::atomic<uint32_t> _refcnt{0};
+
+public:
+    void add_ref() { ++_refcnt; }
+    void del_ref()  { if (--_refcnt == 0) delete this; }
+};
+
+#define IF_RET(e) if (_node) return e; \
+                        else return {};
+
+// the interface for users
+struct NodeWrapper {
+    Node* _node = nullptr;
+    NodeWrapper() = default;
+    NodeWrapper(Node* node) {
+        _node = node;
+        _node->root()->add_ref();
+    }
+    NodeWrapper(const NodeWrapper& rhs) :
+        NodeWrapper(rhs._node) { }
+    NodeWrapper(NodeWrapper&& rhs) {
+        _node = rhs._node;
+        rhs._node = nullptr;
+    }
+    NodeWrapper& operator = (const NodeWrapper& rhs) {
+        auto rt = root_node();
+        auto rrt = rhs.root_node();
+        if (rt != rrt) {
+            if (rt) rt->del_ref();
+            if (rrt) rrt->add_ref();
+        }
+        _node = rhs._node;
+        return *this;
+    }
+    NodeWrapper& operator = (NodeWrapper&& rhs) {
+        if (_node)
+            _node->root()->del_ref();
+        _node = rhs._node;
+        rhs._node = nullptr;
+        return *this;
+    }
+    ~NodeWrapper() {
+        _node->root()->del_ref();
+    }
+    NodeWrapper root() const       { IF_RET(_node->root()); }
+    RootNode* root_node() const    { IF_RET(_node->root()); }
+    NodeWrapper parent() const     { IF_RET(_node->_parent); }
+    NodeWrapper next() const       { IF_RET(_node->_next); }
+    rstring_view32 rkey() const    { IF_RET(_node->_key); }
+    rstring_view32 rvalue() const  { IF_RET(_node->_value); }
+    str key(const char* b) const   { IF_RET(b | rkey()); }
+    str value(const char* b) const { IF_RET(b | rvalue()); }
+    const char* text_begin() const {
+        IF_RET(root()._node->_text_begin);
+    }
+    str key() const    { IF_RET(text_begin() | rkey()); }
+    str value() const  { IF_RET(text_begin() | rvalue()); }
+    bool valid() const { return _node; }
+    operator bool() const { return _node; }
+
+    size_t num_children() const {
+        return _node ? _node->num_children() : 0;
+    }
+    NodeWrapper get(size_t i) const {
+        IF_RET({_node->get(i)});
+    }
+    NodeWrapper get(str key) const {
```

Comment:
What is the result if the get operation fails?

Reply:
an NodeWrapper with nullptr pointer to node implementation

Reply:
This requires a clear explanation and instructions for use.